kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/FileBasedKieServerTemplateStorage.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+package org.kie.server.controller.impl.storage;
+
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.kie.server.api.marshalling.Marshaller;
+import org.kie.server.api.marshalling.MarshallerFactory;
+import org.kie.server.api.marshalling.MarshallingFormat;
+import org.kie.server.api.marshalling.xstream.XStreamMarshaller;
+import org.kie.server.controller.api.model.spec.ServerTemplate;
+import org.kie.server.controller.api.model.spec.ServerTemplateKey;
+import org.kie.server.controller.api.storage.KieServerTemplateStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileBasedKieServerTemplateStorage implements KieServerTemplateStorage {
+    public static final String SERVER_TEMPLATE_FILE_NAME_PROP = "org.kie.server.controller.templatefile";
+    public static final String DEFAULT_SERVER_TEMPLATE_FILENAME = System.getProperty("java.io.tmpdir")+
+    		System.getProperty("file.separator")
+    		+"template_store.xml";
+    private static Logger logger = LoggerFactory.getLogger(FileBasedKieServerTemplateStorage.class);
+    private Map<String, ServerTemplate> templateMap = new ConcurrentHashMap<>();
+    private Map<String, ServerTemplateKey> templateKeyMap = new ConcurrentHashMap<>();
```

Comment:
There is a method that returns a list of ServerTemplateKey objects, and it is faster to grab the keys from the map than it would be to generate them each time. We could store the ServerTemplateKey objects in a List, instead of the Map, but I don't think it would save us much. I can add a method to retrieve a ServerTemplateKey, based on a passed in id. That would at least give a reason for the map to exist.